src/security/forward.h

```diff
@@ -10,7 +10,10 @@
 #define SQUID_SRC_SECURITY_FORWARD_H
 
 #include "base/CbDataList.h"
+#include "security/CertAdaptAlgorithm.h"
```

Comment:
We cannot use unscoped enumeration because these are complex un-named and typedef'd enumerations (as required by the enum strings script) which results in:

>  error: conflicting declaration ‘typedef enum Security::CertSignAlgorithm Security::CertSignAlgorithm’

or 

> error: underlying type mismatch in enum

...depending on whether we name the enumeration (required by the pre-define syntax) or not.

I am now looking into having all caller code include the relevant algorithm header instead of through security/forward.h.

Reply:
okay, done with detatched header includes.

Reply:
This change request has not been fully addressed yet -- `security/CertAdaptAlgorithm.h` is still `#included` in `forward.h`. AFAICT, the only remaining use case is the CertAdaptAlgorithm enumeration. It is simple and can be forward-declared using C++11 unscoped enums. Your response above implies that you cannot forward-declare this enum because our `mk-string-arrays.awk` script requires typedef'd enumerations. However, that script appears to handle many non-typedef enums already, including at least some C++11 enums, so there appears to be no such firm requirement:

```
/^typedef/              { codeSkip = 0; next }
/^enum class / {
  codeSkip = 0
  type = $3
  next;
}
/^enum / {
  codeSkip = 0
  type = $2
  next;
}
codeSkip == 1           { next }
```

Please double check whether a typedef is indeed required and, if there is no simple way to make `mk-string-arrays.awk` work with unscoped typed enums, then please detail the problem.

Reply:
Managed to get this done with the C++11 types enum predefines. Without that there were compile errors from default constructors needing enum type declarations.

Reply:
Good progress, thank you! I see security/DigestAlgorithm.h included by security/forward.h. I am guessing that you have left that inclusion because DigestAlgorithm.h does not define any new "heavy" APIs -- it is essentially a forward-declaration file itself. I can appreciate that point of view, but it would be trivial for somebody to add a heavy declaration to DigestAlgorithm.h without noticing that it is being automatically forwarded to virtually every user of other security APIs.

I see two ways to address this concern:

1. Do not include DigestAlgorithm.h in forward.h. Include DigestAlgorithm.h explicitly where it is needed. This assumes that forward.h itself does not need DigestAlgorithm.h or can be adjusted not to need it.
2. Move DigestAlgorithm.h contents into forward.h. Remove DigestAlgorithm.h.

I am not sure, but option 1 may be slightly better because DigestAlgorithm.h does include some potentially "heavy" OpenSSL and GnuTLS headers.

Reply:
Done in 23504e287